### DISCUSSION TITLE
Getting 401 UNAUTHROIZED when pushing image, even login succeeded

### DISCUSSION THREAD

**Discussion by tigerinus** [quay/quay#1411] [Q&A] [answered]
Hello,

My team is looking to replace Harbor with Project Quay. We've followed the steps at https://github.com/quay/quay/blob/master/docs/quick-local-deployment.md and started our quay instance successfully.

However, when pushing image to this newly setup registry, we got inconsistent messages - On one side it says login successful, but on the other hand it says pushing image failed due to `401 UNAUTHROIZED`.

```bash
tiger@ubuntu-6:/opt/quay$ sudo nerdctl login --insecure-registry -u admin 192.168.193.104
Enter Password:
WARN[0004] skipping verifying HTTPS certs for "192.168.193.104"
WARNING: Your password will be stored unencrypted in /root/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded
tiger@ubuntu-6:/opt/quay$ sudo nerdctl tag hello-world 192.168.193.104/newpathfly/hello-world
tiger@ubuntu-6:/opt/quay$ sudo nerdctl push 192.168.193.104/newpathfly/hello-world --insecure-registry
INFO[0000] pushing as a reduced-platform image (application/vnd.docker.distribution.manifest.list.v2+json, sha256:97a0cf1cfb6b42e8b05a659728699abd3359967853b09667d53d248db40cf647)
WARN[0000] skipping verifying HTTPS certs for "192.168.193.104"
index-sha256:97a0cf1cfb6b42e8b05a659728699abd3359967853b09667d53d248db40cf647:    waiting        |--------------------------------------|
manifest-sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4: waiting        |--------------------------------------|
config-sha256:feb5d9fea6a5e9606aa995e879d862b825965ba48de054caab5ef356dc6b3412:   waiting        |--------------------------------------|
elapsed: 0.1 s                                                                    total:   0.0 B (0.0 B/s)
WARN[0000] server "192.168.193.104" does not seem to support HTTPS, falling back to plain HTTP  error="failed to do request: Head \"https://192.168.193.104/v2/newpathfly/hello-world/blobs/sha256:2db29710123e3e53a794f2694094b9b4338aa9ee5c40b930cb8063a1be392c54\": dial tcp 192.168.193.104:443: connect: connection refused"
index-sha256:97a0cf1cfb6b42e8b05a659728699abd3359967853b09667d53d248db40cf647:    waiting        |--------------------------------------|
manifest-sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4: waiting        |--------------------------------------|
config-sha256:feb5d9fea6a5e9606aa995e879d862b825965ba48de054caab5ef356dc6b3412:   waiting        |--------------------------------------|
elapsed: 0.6 s                                                                    total:   0.0 B (0.0 B/s)
FATA[0000] failed commit on ref "layer-sha256:2db29710123e3e53a794f2694094b9b4338aa9ee5c40b930cb8063a1be392c54": unexpected status from PUT request to http://ubuntu-6/v2/newpathfly/hello-world/blobs/uploads/9927b44c-b355-4ffc-92bb-012b9eb351b9?digest=sha256%3A2db29710123e3e53a794f2694094b9b4338aa9ee5c40b930cb8063a1be392c54: 401 UNAUTHORIZED
```

![image](https://user-images.githubusercontent.com/7172560/175394720-0f17dfb3-19f9-466f-8ded-d36a4efddb69.png)

Also tried with `docker` but got similar error:

```bash
docker push 192.168.193.104/newpathfly/hello-world
Using default tag: latest
The push refers to repository [192.168.193.104/newpathfly/hello-world]
e07ee1baac5f: Pushing [==================================================>]  14.85kB
unauthorized: access to the requested resource is not authorized
```

Please help. Thanks!

**Comment by BillDett** [comment]
Does user 'admin' have write access to the namespace 'newpathfly'? Even if 'admin' is defined as a super user in the Quay config it does not have default write permission on all namespaces.

**Reply by tigerinus** [reply]
Yes `admin` is in the `owner` group of `newpathfly`. I also explicity give `admin` write permission to `newpathfly` but it doesn't change anything.

**Comment by ibazulic** [comment, marked as answer]
There are several possibilities. With what Bill said, how did you set `SERVER_HOSTNAME` variable in Quay's `config.yaml` file? If it's not set to the IP address, where you're pushing, logging in will succeed but push towards any repository will not because the JWT token has an `aud` parameter which is checked explicitly by the auth and compared to the `SERVER_HOSTNAME`. Incidentally, UI login will work because it uses a different auth path than CLI.

**Reply by tigerinus** [reply]
You are actually right. After I set the `SERVER_HOSTNAME` to the same IP address, it all works.

Now my question is - As this registry is only internal facing and runs deep inside our private network, is there anyway to disable authentication? 

The reason behind this is - This server runs on an instance whose IP address for team to access can change from time to time. It's a bit inconvenient to update the config when that happens.

**Reply by ibazulic** [reply]
Glad that one mystery is resolved :-) Unfortunately, you cannot turn off auth in Quay, that's simply not possible. Why not simply use a FQDN  and then leave everything else to DNS, instead of an IP?

**Reply by tigerinus** [reply]
The host has two different IPs, one assigned by cloud network, one assigned by VPN. Depends on how team members access the server, FQDN is resolved to different IP, and it looks like Quay would still returns authorization error (but different message).
